atomics/include/desul/atomics/HIP.hpp

```diff
@@ -261,6 +261,77 @@ __device__ inline
     atomic_fetch_or(T* dest, T val, MemoryOrder, MemoryScopeCore) {
   return atomic_fetch_or(dest, val, MemoryOrder(), MemoryScopeDevice());
 }
+
+}
+
+#define DESUL_HIP_GCC_INTEGRAL_OP_ATOMICS_COMPATIBILITY(MEMORY_ORDER, MEMORY_SCOPE)                 \
+  template <typename T>                                                           \
+  __device__ typename std::enable_if<std::is_integral<T>::value && !Impl::is_hip_atomic_add_type<T>::value, T>::type atomic_fetch_add(  \
+      T* const dest, T value, MEMORY_ORDER, MEMORY_SCOPE) {                       \
+       return Impl::atomic_fetch_oper(Impl::AddOper<T, const T>(), dest, value, MEMORY_ORDER(), MEMORY_SCOPE()); \
```

Comment:
Did you consider have another macro that takes an operation and defines `atomic_fetch_oper` and `atomic_oper_fetch` and call it from your macro?  This would reduce the amount of boiler plate code further.

Reply:
Much as doing this with an X-macro would probably squish it down a good bit, I'm not sure I would do it right now.  Why is this here though, is it falling back from using the hip calls to the gcc builtins on hip-clang? If so, why not just always use the gcc builtins on hip, or go straight for the OpenCL/sycl support builtins so we have the full scoped set?

Reply:
so basically there is some issue with ambiguous calls made more complicated by the fact that I have __device__ and __host__ only functions for in principle different overload sets. The fully generic fallbacks are __host__ __device__, but the GCC host ones are valid for all std::is_integral, while the HIP native ones are only there for the set which is Impl::is_hip_atomic_add_type. So I needed set of __device__ only equivalents to the __host__ only equivalents provided by GCC.

Reply:
Ah, that makes good sense, thanks!